### DISCUSSION TITLE
Improving binary data handling functions

### DISCUSSION THREAD

**Discussion by coyo-t** [godotengine/godot-proposals#5558] [Scripting]
As someone that works a lot with binary formats, I find Godot's `File.*` (or `FileAccess.*` in GD 4?) functions incredibly lacking and janky (especially coming from Gamemaker, where the [`buffer_*`](https://manual.yoyogames.com/#t=GameMaker_Language%2FGML_Reference%2FBuffers%2FBuffers.htm) functions are very nearly perfect to me, though are lacking in some text-oriented functions).

Originally I had a large list of points about `FileAccess`'s methods and how bizzare and disjointed they are compared to `buffer_*`'s, but then I took a look at `PackedByteArray`'s methods and hey! It's fairly synonymous with GM's `buffer_*` functions! Except it lacks the cursor capabilities of `buffer` and `FileAccess`. And both `FileAccess` and `PackedByteArray` lack the cursor handling modes and alignment abilities `buffer` has to offer (IE `buffer_fixed/grow/wrap`).

So I want to look into what can be done about this. Ideally, I'd just have GM's `buffer` ported over near-verbatim into GD, with the added feat that it's general purpose (IE can work with `PackedByteArray` or file streams, of which `buffer` can only do the GM equivalent of the former). However that's probably an undertaking and a half, so it'd probably just be easier to make the two (FA and PBA) on-par with each other and *eventually* that could be condensed into one generic `Buffer` system.
Or maybe just port `buffer` over verbatim but have it *not* be a generic takeover of FA and PBA's encoding/decoding abilities, and just have it be its own thing that can be dumped to a PBA. I think that one's the easiest of the 3, but I'd like to know from someone more familiar with GD's internals than I am.

**Comment by Zireael07** [comment]
Drive by comment: 
1) I wanted to use binary serialization but without using File (just for in-memory storage). The fact you can't do this is NOT called out in the Binary serialization docs page...
2) Buffers or strings of bytes or whatever you call them should have a function to convert them to int (for what is called "packing" stuff in integers)
